### PR TITLE
Save CodePrinter in Codegen to use information collected during printing stage

### LINKED ISSUE #483: Imports are included in both the .h and .c file
**Describe the bug**
Currently most imports are placed in the `.h` files. Normally the number of imports in the `.h` file should be minimised. So only the files necessary for the signatures should be imported. It may be difficult to tell which files are required so currently everything is imported in the `.h` file. The `.c` file then imports the `.h` file and any discovered imports (e.g. `math.h`, `complex.h`). These discovered imports may exist in both the `.h` and the `.c` file. This is notably the case for any function with a complex argument

**Language**
C

### PR DESCRIPTION
**Pull request by rakati**
Change the way we call the codeprinter to be saved in Codegen:

- Change the printer_register to map languages to its codeprinter;
- Call directly the codeprinter without using the intermediate function;
- Save the codeprinter in _printer to use its information;
- Print moduleheader first to include only libraries used by it.

Closes #483

### CODE REVIEW

File: pyccel/codegen/codegen.py [hunk 1]
```diff
@@ -149,6 +154,23 @@ def code(self):
...
+        """ Set the current codeprinter instance"""
+        # Get language used (default language used is fortran)
+        language = settings.pop('language', 'fortran')
+
+        # Set language
+        if not language in ['fortran', 'c', 'python']:
+            raise ValueError('{} language is not available'.format(language))
+        self._language = language
+
+        # instanciat codePrinter
```
Comment (EmilyBourne) [inline pyccel/codegen/codegen.py:167]: typo `instanciat`->`instantiate`

File: pyccel/codegen/printing/ccode.py [hunk 2]
```diff
@@ -274,7 +274,7 @@ def _print_ModuleHeader(self, expr):
         funcs = '\n\n'.join('{};'.format(self.function_signature(f)) for f in expr.module.funcs)
 
         # Print imports last to be sure that all additional_imports have been collected
-        imports = [*expr.module.imports, *map(Import, self._additional_imports)]
+        imports = [*expr.module.imports,*map(Import, self._additional_imports)]
```
Comment (EmilyBourne) [inline pyccel/codegen/printing/ccode.py:277]: Not sure why you changed this file

File: pyccel/codegen/printing/pycode.py [hunk 3]
```diff
@@ -34,9 +34,9 @@ class PythonCodePrinter(SympyPythonCodePrinter):
     ))
     _kc = {k: ''+v for k, v in _known_constants_math.items()}
 
-    def __init__(self, settings=None):
+    def __init__(self, parser=None, settings={}):
```
Comment (EmilyBourne) [inline pyccel/codegen/printing/pycode.py:37]: The default is None to make codacy happy. See here for why : https://stackoverflow.com/a/41686973/5957195

Comment (yguclu) [reply pyccel/codegen/printing/pycode.py:37]: As a rule of thumb, default values should be immutable.

File: pyccel/codegen/printing/ccode.py (hunk 2 above)
Comment (rakati) [reply pyccel/codegen/printing/ccode.py:277]: this was revert changes.
if you follow the commits you will see that.

Comment (EmilyBourne) [reply pyccel/codegen/printing/ccode.py:277]: Ah so you changed the file then changed it back and the only difference left is the removed space. In that case please checkout this file from the master branch so it is clear that it is unchanged

**Review by EmilyBourne — APPROVED** [review]

**Review by yguclu — APPROVED** [review]

**Pull request merged**